assets/js/googlesitekit/data/create-fetch-store.test.js

```diff
@@ -0,0 +1,328 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { createRegistry } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import {
+	muteConsole,
+	subscribeUntil,
+	unsubscribeFromAll,
+} from 'tests/js/utils';
+import { createFetchStore } from './create-fetch-store';
+
+const STORE_NAME = 'test/some-data';
+const STORE_PARAMS = {
+	baseName: 'getSomeData',
+	controlCallback: ( params ) => {
+		const { aParam, objParam } = params;
+		return API.get( 'core', 'test', 'some-data', {
+			aParam,
+			objParam,
+		} );
+	},
+	reducerCallback: ( state, response, params ) => {
+		const { aParam } = params;
+		return {
+			...state,
+			data: {
+				...( state.data || {} ),
+				[ aParam ]: response,
+			},
+		};
+	},
+	argsToParams: ( objParam, aParam ) => {
+		invariant( 'object' === typeof objParam, 'objParam is required.' );
+		invariant( 'undefined' !== typeof aParam, 'aParam is required.' );
+		return {
+			objParam,
+			aParam,
+		};
+	},
+};
+
+describe( 'createFetchStore store', () => {
+	let apiFetchSpy;
+	let dispatch;
+	let registry;
+	let select;
+	let storeDefinition;
+	let store;
+
+	beforeAll( () => {
+		API.setUsingCache( false );
+	} );
+
+	beforeEach( () => {
+		registry = createRegistry();
+
+		storeDefinition = createFetchStore( STORE_PARAMS );
+		registry.registerStore( STORE_NAME, storeDefinition );
+		dispatch = registry.dispatch( STORE_NAME );
+		store = registry.stores[ STORE_NAME ].store;
+		select = registry.select( STORE_NAME );
+
+		apiFetchSpy = jest.spyOn( { apiFetch }, 'apiFetch' );
+	} );
+
+	afterAll( () => {
+		API.setUsingCache( true );
+	} );
+
+	afterEach( () => {
+		unsubscribeFromAll( registry );
+		apiFetchSpy.mockRestore();
+	} );
+
+	describe( 'actions', () => {
+		it( 'includes the expected actions', () => {
+			const fetchStoreDefinition = createFetchStore( {
+				baseName: 'SaveSomeData',
+				controlCallback: async () => true,
+				reducerCallback: ( state ) => state,
+			} );
+
+			expect( Object.keys( fetchStoreDefinition.actions ) ).toEqual( [
+				'fetchSaveSomeData',
+				'receiveSaveSomeData',
+			] );
+		} );
+
+		describe( 'fetch', () => {
+			it( 'validates parameters based on argsToParams', () => {
+				const consoleErrorSpy = jest.spyOn( global.console, 'error' );
+
+				dispatch.fetchGetSomeData();
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'objParam is required.' );
+
+				dispatch.fetchGetSomeData( 123 );
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'objParam is required.' );
+
+				dispatch.fetchGetSomeData( {} );
+				expect( consoleErrorSpy ).toHaveBeenCalledWith( 'aParam is required.' );
+
+				consoleErrorSpy.mockClear();
```

Comment:
Make sure to restore rather than clear to ensure it's back to it's normal state for other tests.
```suggestion
				consoleErrorSpy.mockRestore();
```
Also, I think this kind of cleanup should always happen in an `afterEach` or another lifecycle callback as an error in the test would otherwise cause the cleanup to be skipped and cause problems in subsequent tests.

Reply:
Using `mockRestore` here doesn't work, it will result in errors elsewhere:
```
    TypeError: global.console.error.mockClear is not a function

      1 | beforeEach( () => {
    > 2 | 	global.console.error.mockClear();
        | 	                     ^
      3 | 	global.console.warn.mockClear();
      4 | 	global.console.log.mockClear();
      5 | 	global.console.info.mockClear();

      at Object.mockClear (tests/js/setup-before-after.js:2:23)
```

I think we can leave this as is for now and look into it separately, there are multiple places in our existing code place where this approach is used.

Reply:
Oh okay, thanks for clarifying. It should still be called in a lifecycle callback though rather than just at the end of the test.